File: kavya_app/lib/services/auth_service.dart
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../core/router/app_router.dart';
import '../models/user.dart';
import '../providers/auth_provider.dart';
import 'api_service.dart';

final authServiceProvider = Provider((ref) => AuthService(ref));

class AuthService {
  final Ref _ref;
  final FlutterSecureStorage _storage = const FlutterSecureStorage();
  final ApiService _apiService = ApiService();

  AuthService(this._ref);

  /// Step-1: request OTP. Returns a map containing:
  ///   { 'otp_required': true, 'session_id': '...', 'phone_masked': '...' }
  /// For admin/direct login returns:
  ///   { 'otp_required': false }
  Future<Map<String, dynamic>> loginStep1(String identifier, String password) async {
    // Detect phone vs email/employee-ID
    final isPhone = RegExp(r'^\d{10}$').hasMatch(identifier.trim());
    if (isPhone) {
      // Phone-based: call send-otp (triggers 2Factor SMS)
      final resp = await _apiService.sendOtp(identifier.trim(), password);
      final data = resp['data'] as Map<String, dynamic>? ?? {};
      return {
        'otp_required': true,
        'session_id': data['session_id'],
        'phone_masked': data['phone_masked'],
        'delivery': data['delivery'],
      };
    } else {
      // Email / employee-ID: direct login (admin/staff without phone OTP)
      final response = await _apiService.login(identifier, password);
      final data = response['data'] as Map<String, dynamic>? ?? response;
      await _saveTokensAndNavigate(data);
      return {'otp_required': false};
    }
  }

  /// Step-2: verify the 6-digit OTP entered by the user.
  Future<void> confirmOtp(String sessionId, String otp) async {
    final response = await _apiService.verifyOtp(sessionId, otp);
    final data = response['data'] as Map<String, dynamic>? ?? response;
    await _saveTokensAndNavigate(data);
  }

  Future<void> login(String email, String password) async {
    final response = await _apiService.login(email, password);
    final data = response['data'] as Map<String, dynamic>? ?? response;
    await _saveTokensAndNavigate(data);
  }

  /// Phone OTP login for market (hired) drivers.
  /// [sessionId], [accessToken], [otpCode] come from the OTP flow.
  Future<void> loginMarketDriver({
    required String sessionId,
    required String accessToken,
    required String otpCode,
  }) async {
    final response = await _apiService.post('/auth/market-driver/verify-otp', data: {
      'session_id': sessionId,
      'access_token': accessToken,
      'otp_code': otpCode,
    });
    final data = response['data'] as Map<String, dynamic>? ?? response;
    await _saveTokensAndNavigate(data);
  }

  Future<void> _saveTokensAndNavigate(Map<String, dynamic> data) async {
    final token = data['access_token'] as String?;
    final refreshToken = data['refresh_token'] as String?;
    final user = data['user'] as Map<String, dynamic>?;

    if (token == null || user == null) {
      throw Exception('Invalid login response');
    }

    final roles = user['roles'] as List<dynamic>? ?? [];
    final primaryRole = roles.isNotEmpty ? roles[0].toString() : 'unknown';
    // For market drivers name comes from backend user.name field directly
    final phoneRaw = user['phone']?.toString() ?? '';
    final userName = primaryRole == 'market_driver'
        ? (user['name']?.toString() ?? 'Driver')
        : '${user['first_name'] ?? ''} ${user['last_name'] ?? ''}'.trim();
    final resolvedName = userName.isNotEmpty ? userName : user['email']?.toString() ?? 'User';

    await _storage.write(key: 'access_token', value: token);
    if (refreshToken != null) {
      await _storage.write(key: 'refresh_token', value: refreshToken);
    }
    await _storage.write(key: 'primary_role', value: primaryRole);
    await _storage.write(key: 'user_name', value: resolvedName);
    await _storage.write(key: 'user_id', value: user['id'].toString());
    await _storage.write(key: 'user_email', value: user['email']?.toString() ?? '');
    await _storage.write(key: 'user_phone', value: user['phone']?.toString() ?? phoneRaw);
    await _storage.write(key: 'user_is_active', value: (user['is_active'] as bool? ?? true).toString());
    await _storage.write(key: 'user_avatar_url', value: user['avatar_url']?.toString() ?? '');

    _ref.read(authProvider.notifier).setUser(User(
      id: user['id'].toString(),
      name: resolvedName,
      email: user['email']?.toString() ?? '',
      roles: roles.map((r) => r.toString()).toList(),
      phone: user['phone']?.toString(),
      isActive: user['is_active'] as bool? ?? true,
      avatarUrl: user['avatar_url']?.toString(),
    ));

    _navigateForRole(primaryRole);
  }

  void _navigateForRole(String role) {
    final router = _ref.read(routerProvider);
    switch (role) {
      case 'driver':
        router.go('/driver/today');
        break;
      case 'fleet_manager':
        router.go('/fleet/home');
        break;
      case 'accountant':
        router.go('/accountant/home');
        break;
      case 'finance_manager':
      case 'FINANCE_MANAGER':
        router.go('/finance/home');
        break;
      case 'project_associate':
        router.go('/pa/dashboard');
        break;
      case 'admin':
      case 'super_admin':
        router.go('/admin/dashboard');
        break;
      case 'pump_operator':
        router.go('/pump/home');
        break;
      case 'branch_manager':
        router.go('/branch/home');
        break;
      case 'manager':
        router.go('/manager/dashboard');
        break;
      case 'market_driver':
        router.go('/market-driver/trips');
        break;
      case 'tyre_inspector':
        router.go('/tyre/dashboard');
        break;
      default:
        router.go('/web-only');
    }
  }

  Future<void> logout() async { // [cite: 41]
    try {
      // Optional: Call POST /auth/logout if implemented on backend
      // await _apiService.logout(); 
    } catch (_) {} 
    
    await _storage.deleteAll();
    _ref.read(routerProvider).go('/login');
  }
}

File: frontend/src/services/authService.ts
import api from './api';
import type { LoginRequest, LoginResponse, RegisterRequest, User } from '@/types';

const unwrap = <T = any>(payload: any): T => (payload?.data ?? payload) as T;

export const authService = {
  login: async (credentials: LoginRequest): Promise<LoginResponse> => {
    const data = await api.post<LoginResponse>('/auth/login', {
      identifier: credentials.identifier,
      password: credentials.password,
    });
    return unwrap(data);
  },

  register: async (payload: RegisterRequest): Promise<User> => {
    const data = await api.post<User>('/auth/register', payload);
    return data;
  },

  getMe: async (): Promise<User> => {
    const data = await api.get<User>('/auth/me');
    return unwrap(data);
  },

  refreshToken: async (refreshToken: string): Promise<LoginResponse> => {
    const data = await api.post<LoginResponse>('/auth/refresh', { refresh_token: refreshToken });
    return unwrap(data);
  },

  changePassword: async (currentPassword: string, newPassword: string): Promise<void> => {
    await api.post('/auth/change-password', {
      current_password: currentPassword,
      new_password: newPassword,
    });
  },

  forgotPassword: async (email: string): Promise<void> => {
    await api.post('/auth/forgot-password', { email });
  },

  resetPassword: async (token: string, newPassword: string): Promise<void> => {
    await api.post('/auth/reset-password', { token, new_password: newPassword });
  },

  logout: async (): Promise<void> => {
    const refreshToken = localStorage.getItem('refresh_token');
    try {
      await api.post('/auth/logout', refreshToken ? { refresh_token: refreshToken } : {});
    } catch { /* ignore */ }
    sessionStorage.removeItem('access_token');
    sessionStorage.removeItem('portal_token');
    sessionStorage.removeItem('portal_role');
    sessionStorage.removeItem('portal_name');
    localStorage.removeItem('access_token');
    localStorage.removeItem('refresh_token');
  },

  sendOtp: async (
    credential: { phone?: string; identifier?: string; channel?: string },
    password: string,
  ): Promise<{ session_id: string; phone_masked?: string; email_masked?: string; delivery?: string }> => {
    const data = await api.post<{ session_id: string; phone_masked?: string; email_masked?: string; delivery?: string }>(
      '/auth/send-otp',
      { ...credential, password, channel: credential.channel ?? 'sms' },
    );
    return unwrap(data);
  },

  verifyOtp: async (session_id: string, otp: string): Promise<LoginResponse> => {
    const data = await api.post<LoginResponse>('/auth/verify-otp', { session_id, otp });
    return unwrap(data);
  },
};
